### ISSUE TITLE
Bug: no-use-before-define reports false positives

### ISSUE THREAD

**Issue by mareolan** [eslint/eslint#17003]
### Environment

Node version: 16.14.2
npm version: 8.19.4
Local ESLint version: 8.36.0
Global ESLint version: -
Operating System: Windows 10

### What parser are you using?

@babel/eslint-parser

### What did you do?

<details>
<summary>Configuration</summary>

```
{
  ...,
  rules: {
    "no-use-before-define": ["error", { functions: false }],
  }
}
```
</details>

I expected that the rule would report only real or potential problems regarding ES6 const/let usage ("Uncaught ReferenceError: Cannot access 'b' before initialization"), but instead the rule reports many false positives. And if it is relaxed using e.g. { variables: false } then it also stops detecting certain real problems. I'd like the rule to behave like this:
```js
// E1 - this should be ok (function a() wasn't called before 'b' gets declared)
// currently it is (incorrectly) reported as: 'b' was used before it was defined.
const a = () => console.log(b);
const b = 10;

// E2 - this should be reported (problem) because function is called before 'b' gets declared
const a = () => console.log(b);
a(); // this causes the usage of 'b' above not to be ok
const b = 10;

// E3 - this should be reported (potential problem), maybe configurable
const a = someFn(() => {
  console.log(b); // someFn could have executed this arrow function immediately, i.e. 'b' wouldn't be declared yet
});
const b = 10;
```

Basically, if a function uses variable from outer scope, it should report the problem with such variable only if the function was really (E2) / potentially (E3) called.

Alternatively, if the rule is currently intended for something like visual order of declarations (i.e. not primarily supposed to detect real problems), then it would be nice to create another rule that detects the real/potential problems.

### What did you expect to happen?

Only real and potential problems should be reported.

### What actually happened?

Either there were false positives, or if the configuration was relaxed then there were omitted real problems.

### Participation

- [ ] I am willing to submit a pull request for this issue.

**Comment by NotWearingPants** [comment]
To what extent do you expect the static analysis to detect such cases? Would you expect the following to error or not?
```js
const a = () => console.log(b);
a.call();
const b = 10;
```
```js
const a = () => console.log(b);
a.toString();
const b = 10;
```
```js
const a = () => console.log(b);
if (false) { a(); }
const b = 10;
```
```js
let a = () => console.log(b);
a = () => console.log('Phew');
a();
const b = 10;
```
```js
const c = () => a();
const a = () => console.log(b);
c();
const b = 10;
```
```js
const a = () => console.log(b);
const c = { get foo() { a(); } };
c['f' + 'o' + 'o'];
const b = 10;
```

More specifically my question is, can you define "potential problem"? What is an acceptable false-positive and what isn't?

**Comment by Rec0iL99** [comment]
Hi @mareolan, thanks for the issue and thanks for reaching out @mareolan.

```js
const a = () => console.log(b);
const b = 10;
```

`no-use-before-define` rule falls under the title *Possible Problems* in the docs for [rules reference](https://eslint.org/docs/latest/rules/#possible-problems). I think this is why the rule reports problems even in places where the runtime is currently not affected but could be affected in the future. ESLint is basically warning the user that there could be possible logic errors in the code.

The `no-use-before-define` rule is partially stylistic and intentionally more opinionated (ref https://github.com/eslint/eslint/issues/11903#issuecomment-505469231). This is why it's **not** included in `eslint:recommended` because people have different stylistic preferences in their codebases (ref #11903).

This throws `ReferenceError` during runtime:

```js
const a = () => console.log(b);
a();
const b = 10;

// Expected output: ReferenceError: Cannot access 'b' before initialization
```

but not this:

```js
const a = () => console.log(b);
const b = 10;
a();

// Expected output: 10
```

ESLint reports the rule in both places. The code samples do the same thing but different output.

Quoting the docs here, "....this can be confusing and some believe it is best to always declare variables and functions before using them.". So I feel it's more of a preference for the user to use the rule or not.

> Alternatively, if the rule is currently intended for something like visual order of declarations (i.e. not primarily supposed to detect real problems), then it would be nice to create another rule that detects the real/potential problems.

I'm leaving this open for the @eslint/eslint-tsc to decide if a **new rule** as the author suggested is needed here. @mareolan if approved you would need to create a new rule proposal to move forward. Thanks.
